### PR TITLE
Move to Microsoft.Extensions.DependencyInjection

### PR DESCRIPTION
**Pull request by Bond-009**
This PR replaces SimpleInjector with Microsoft.Extensions.DependencyInjection.

### CODE REVIEW

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 1]
```diff
@@ -748,107 +658,107 @@ protected virtual IHttpClient CreateHttpClient()
...
 
-            RegisterSingleInstance(JsonSerializer);
 
-            RegisterSingleInstance(LoggerFactory, false);
-            RegisterSingleInstance(Logger);
+            serviceCollection.AddSingleton(JsonSerializer);
 
-            RegisterSingleInstance(EnvironmentInfo);
+            serviceCollection.AddSingleton(LoggerFactory);
+            serviceCollection.AddLogging();
```
Comment (wtayl0r) [inline Emby.Server.Implementations/ApplicationHost.cs]: Does this registration add anything we need specifically? we've already registered the factory and the logger.

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 2]
```diff
@@ -748,107 +658,107 @@ protected virtual IHttpClient CreateHttpClient()
...
-            RegisterSingleInstance(ZipClient);
+            serviceCollection.AddSingleton(ZipClient);
 
             HttpResultFactory = new HttpResultFactory(LoggerFactory, FileSystemManager, JsonSerializer, CreateBrotliCompressor());
-            RegisterSingleInstance(HttpResultFactory);
+            serviceCollection.AddSingleton(HttpResultFactory);
 
-            RegisterSingleInstance<IServerApplicationHost>(this);
-            RegisterSingleInstance<IServerApplicationPaths>(ApplicationPaths);
+            serviceCollection.AddSingleton<IServerApplicationHost>(this);
```
Comment (wtayl0r) [inline Emby.Server.Implementations/ApplicationHost.cs]: Would move the IServerApplicationHost registration up to the top with the IApplicationHost as they are registering different interfaces against the same object.

File: Emby.Server.Implementations/ApplicationHost.cs (hunk 1 above)
Comment (Bond-009) [reply Emby.Server.Implementations/ApplicationHost.cs]: Yes it adds support for ‘Logger<T>’

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 3]
```diff
@@ -458,138 +452,58 @@ public string SystemId
...
             try
             {
-                return Container.GetInstance(type);
+                return ActivatorUtilities.CreateInstance(_serviceProvider, type);
             }
             catch (Exception ex)
             {
-                Logger.LogError(ex, "Error creating {type}", type.FullName);
+                Logger.LogError(ex, "Error creating {Type}", type);
                 // Don't blow up in release mode
```
Comment (JustAMan) [inline Emby.Server.Implementations/ApplicationHost.cs]: this comment looks strange, if it's only in release mode then next `return null` should be under `#ifdef DEBUG` or something

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 4]
```diff
@@ -863,81 +773,80 @@ protected void RegisterResources()
...
-            RegisterSingleInstance(ImageProcessor);
+            serviceCollection.AddSingleton(ImageProcessor);
 
             TVSeriesManager = new TVSeriesManager(UserManager, UserDataManager, LibraryManager, ServerConfigurationManager);
-            RegisterSingleInstance(TVSeriesManager);
+            serviceCollection.AddSingleton(TVSeriesManager);
 
             var encryptionManager = new EncryptionManager();
-            RegisterSingleInstance<IEncryptionManager>(encryptionManager);
+            serviceCollection.AddSingleton<IEncryptionManager>(encryptionManager);
```
Comment (JustAMan) [inline Emby.Server.Implementations/ApplicationHost.cs]: make one line:
```suggestion
            serviceCollection.AddSingleton<IEncryptionManager>(new EncryptionManager());
```

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 5]
```diff
@@ -863,81 +773,80 @@ protected void RegisterResources()
...
-            RegisterSingleInstance(ChannelManager);
+            serviceCollection.AddSingleton(ChannelManager);
 
             SessionManager = new SessionManager(UserDataManager, LoggerFactory, LibraryManager, UserManager, musicManager, DtoService, ImageProcessor, JsonSerializer, this, HttpClient, AuthenticationRepository, DeviceManager, MediaSourceManager, TimerFactory);
-            RegisterSingleInstance(SessionManager);
+            serviceCollection.AddSingleton(SessionManager);
 
-            var dlnaManager = new DlnaManager(XmlSerializer, FileSystemManager, ApplicationPaths, LoggerFactory, JsonSerializer, this, assemblyInfo);
-            RegisterSingleInstance<IDlnaManager>(dlnaManager);
+            serviceCollection.AddSingleton<IDlnaManager>(new DlnaManager(XmlSerializer, FileSystemManager, ApplicationPaths, LoggerFactory, JsonSerializer, this, assemblyInfo));
```
Comment (JustAMan) [inline Emby.Server.Implementations/ApplicationHost.cs]: and this maybe return to two-line for the sake of readability, or wrap long line

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 6]
```diff
@@ -863,81 +773,80 @@ protected void RegisterResources()
...
             UserViewManager = new UserViewManager(LibraryManager, LocalizationManager, UserManager, ChannelManager, LiveTvManager, ServerConfigurationManager);
-            RegisterSingleInstance(UserViewManager);
+            serviceCollection.AddSingleton(UserViewManager);
 
             NotificationManager = new NotificationManager(LoggerFactory, UserManager, ServerConfigurationManager);
-            RegisterSingleInstance(NotificationManager);
+            serviceCollection.AddSingleton(NotificationManager);
 
-            RegisterSingleInstance<IDeviceDiscovery>(new DeviceDiscovery(LoggerFactory, ServerConfigurationManager, SocketFactory, TimerFactory));
+            serviceCollection.AddSingleton<IDeviceDiscovery>(new DeviceDiscovery(LoggerFactory, ServerConfigurationManager, SocketFactory, TimerFactory));
```
Comment (JustAMan) [inline Emby.Server.Implementations/ApplicationHost.cs]: make two-line or wrap?

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 7]
```diff
@@ -1561,150 +1409,66 @@ public void Restart()
...
+            if (Directory.Exists(ApplicationPaths.PluginsPath))
+            {
+                foreach (var file in Directory.EnumerateFiles(ApplicationPaths.PluginsPath, "*.dll", SearchOption.TopDirectoryOnly))
+                {
+                    Logger.LogInformation("Loading assembly {Path}", file);
+                    yield return Assembly.LoadFile(file);
+                }
+            }
 
             // Gets all plugin assemblies by first reading all bytes of the .dll and calling Assembly.Load against that
```
Comment (JustAMan) [inline Emby.Server.Implementations/ApplicationHost.cs]: what this comment means and is it still relevant?

File: Emby.Server.Implementations/ApplicationHost.cs (hunk 4 above)
Comment (Bond-009) [reply Emby.Server.Implementations/ApplicationHost.cs]: `encryptionManager` is probably used further in the function.

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 8]
```diff
@@ -1279,78 +1190,15 @@ protected void DiscoverTypes()
...
-                var assembly = assemblyInfo.Item1;
-                var path = assemblyInfo.Item2;
-
-                if (path == null)
-                {
-                    Logger.LogInformation("Loading {assemblyName}", assembly.FullName);
-                }
-                else
+            AllConcreteTypes = GetComposablePartAssemblies()
+                .SelectMany(x => x.ExportedTypes)
```
[the remark this answers is not in the document]
Comment (cvium) [reply Emby.Server.Implementations/ApplicationHost.cs]: Where are we on this? Are you still opposed to using DefinedTypes?

File: Emby.Server.Implementations/ApplicationHost.cs (hunk 7 above)
Comment (JustAMan) [reply Emby.Server.Implementations/ApplicationHost.cs]: I think it's no longer relevant as we've switched to `LoadFrom`, right? What about live update of libraries though?

**Pull request merged**